.github/cicd/functions.sh

```diff
@@ -0,0 +1,174 @@
+
+function pipeline_args() {
+    export PIPELINE_ARGS="$*"
+}
+
+function build() {
+    source configurations/$2.conf
+    echo "Building $2 (${APP_NAME}) to $PROJECT_PREFIX-$3..."
+    buildPresent=$(oc get bc/"$APP_NAME-$BRANCH_LOWER" --ignore-not-found=true)
+    if [ -z "${buildPresent}" ];
+    then
+        MODE="apply"
+    else
+        MODE="create"
+    fi;
+    if [ "${BRANCH_LOWER}" == "develop" ] || [ "${BRANCH_LOWER}" == "master" ];
+    then
+        echo "oc process -f ./${TEMPLATE_DIRECTORY}/${BUILD_CONFIG_TEMPLATE} -p NAME=${APP_NAME} -p VERSION=${BUILD_NUMBER} -p SOURCE_CONTEXT_DIR=${SOURCE_CONTEXT_DIR} -p SOURCE_REPOSITORY_URL=${GIT_URL} -p SOURCE_REPOSITORY_REF=${BRANCH_NAME} -p OC_NAMESPACE=$PROJECT_PREFIX -p OC_APP=$3 ${@:4} | oc ${MODE} -f - --namespace=$PROJECT_PREFIX-$3"
+        oc process -f ./"${TEMPLATE_DIRECTORY}/${BUILD_CONFIG_TEMPLATE}" \
+        -p NAME="${APP_NAME}" \
+        -p VERSION="${BUILD_NUMBER}" \
+        -p SOURCE_CONTEXT_DIR="${SOURCE_CONTEXT_DIR}" \
+        -p SOURCE_REPOSITORY_URL="${GIT_URL}" \
+        -p SOURCE_REPOSITORY_REF="${BRANCH_NAME}" \
+        -p OC_NAMESPACE="$PROJECT_PREFIX" \
+        -p OC_APP="$3" ${@:4} --output="yaml" | oc "${MODE}" -f - --namespace="$PROJECT_PREFIX-$3" --overwrite=true --all
+    else
+        echo "oc process -f ./${TEMPLATE_DIRECTORY}/${BUILD_CONFIG_TEMPLATE} -p NAME=${APP_NAME} -p VERSION=${BUILD_NUMBER} -p SUFFIX=-${BRANCH_LOWER} -p SOURCE_CONTEXT_DIR=${SOURCE_CONTEXT_DIR} -p SOURCE_REPOSITORY_URL=${GIT_URL} -p SOURCE_REPOSITORY_REF=${BRANCH_NAME} -p OC_NAMESPACE=$PROJECT_PREFIX -p OC_APP=$3 ${@:4} | oc ${MODE} -f - --namespace=$PROJECT_PREFIX-$3"
+        oc process -f ./"${TEMPLATE_DIRECTORY}/${BUILD_CONFIG_TEMPLATE}" \
+        -p NAME="${APP_NAME}" \
+        -p VERSION="${BUILD_NUMBER}" \
+        -p SUFFIX="-${BRANCH_LOWER}" \
+        -p SOURCE_CONTEXT_DIR="${SOURCE_CONTEXT_DIR}" \
+        -p SOURCE_REPOSITORY_URL="${GIT_URL}" \
+        -p SOURCE_REPOSITORY_REF="${CHANGE_BRANCH}" \
+        -p OC_NAMESPACE="$PROJECT_PREFIX" \
+        -p OC_APP="$3" ${@:4} --output="yaml" | oc "${MODE}" -f - --namespace="$PROJECT_PREFIX-$3" --overwrite=true --all
+    fi;
+    if [ "$BUILD_REQUIRED" == true ];
+    then
+        echo "Building oc start-build $APP_NAME$SUFFIX -n $PROJECT_PREFIX-$3 --wait --follow ..."
+        oc start-build "$APP_NAME$SUFFIX" -n "$PROJECT_PREFIX-$3" --wait --follow
+    else
+        echo "Deployment should be automatic..."
+    fi
+}
+
+function deploy() {
+    source configurations/$2.conf
+    echo "Deploying $2 (${APP_NAME}) to $3 ..."
+    deployPresent=$(oc get dc/"${APP_NAME}-${BRANCH_LOWER}" --ignore-not-found=true)
+    if [ -z "${deployPresent}" ];
+    then
+        MODE="apply"
+    else
+        MODE="create"
+    fi;
+    echo "Interpreted:"
+    if [ "${BRANCH_LOWER}" == "develop" ] || [ "${BRANCH_LOWER}" == "master" ];
+    then
+        oc process -f ./"${TEMPLATE_DIRECTORY}/${DEPLOY_CONFIG_TEMPLATE}" \
```

Comment:
do we need to be thinking about echoing/logging in these if/elses for visibility?

Reply:
We could - would be useful for debugging.  Build has it because that's where things usually go wrong.  We could add to deploy but I'll look into that later.